envpool/sokoban/sokoban_node.cc

```diff
@@ -0,0 +1,196 @@
+        }
+      } else if (is_goal) {
+        std::cout << ".";
+      } else {
+        std::cout << " ";
+      }
+    }
+    std::cout << std::endl;
+  }
+}
+
+SokobanNode* SokobanNode::GetChildNode(int action_idx) {
+  int delta_x = kDelta.at(action_idx).first;
+  int delta_y = kDelta.at(action_idx).second;
+  int new_player_x = player_x + delta_x;
+  int new_player_y = player_y + delta_y;
+  // check if the move is valid
+  if (CheckWall(new_player_x, new_player_y)) {
+    return nullptr;
+  }
+  // check if (new_player_x, new_player_y) is a box, if it is not, return a
+  // new SokobanNode with the new player position
+  std::vector<std::pair<int, int>> new_boxes = boxes;
+  for (size_t i = 0; i < boxes.size(); i++) {
+    if (boxes.at(i).first == new_player_x &&
+        boxes.at(i).second == new_player_y) {
+      int new_box_x = boxes.at(i).first + delta_x;
+      int new_box_y = boxes.at(i).second + delta_y;
+      // check if the box can move
+      if (CheckWall(new_box_x, new_box_y)) {
+        return nullptr;
+      }
+      // check if the box is blocked by another box
+      for (size_t j = 0; j < boxes.size(); j++) {
+        if (boxes.at(j).first == new_box_x && boxes.at(j).second == new_box_y) {
+          return nullptr;
+        }
+      }
+      // update the box position
+      new_boxes.at(i).first = new_box_x;
+      new_boxes.at(i).second = new_box_y;
+      if (delta_y != 0) {
+        std::sort(
+            new_boxes.begin(), new_boxes.end(),
+            [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
+              if (a.second != b.second) {
+                return a.second < b.second;
+              }
+              return a.first < b.first;
+            });
+      }
+      break;
+    }
+  }
+  return new SokobanNode(dim_room, new_player_x, new_player_y, new_boxes, walls,
+                         this, action_idx);
+}
+
+bool SokobanNode::CheckWall(int x, int y) {
+  if (x < 0 || x >= dim_room || y < 0 || y >= dim_room) {
+    return true;
+  }
+  return walls->at(x + y * dim_room);
+}
+
+size_t SokobanNode::Hash() const {
+  size_t hash = 0;
+  hash = (hash * 397) ^ std::hash<int>{}(player_x);
+  hash = (hash * 397) ^ std::hash<int>{}(player_y);
+  for (size_t i = 0; i < boxes.size(); i++) {
+    hash = (hash * 397) ^ std::hash<int>{}(boxes.at(i).first);
+    hash = (hash * 397) ^ std::hash<int>{}(boxes.at(i).second);
+  }
+  return hash;
+}
+
+bool SokobanNode::IsGoal(SokobanNode& goal_node) {
+  for (size_t i = 0; i < boxes.size(); i++) {
+    bool matched = false;
+    for (size_t j = 0; j < goal_node.boxes.size(); j++) {
+      if (boxes.at(i).first == goal_node.boxes.at(j).first &&
+          boxes.at(i).second == goal_node.boxes.at(j).second) {
+        matched = true;
+        break;
+      }
+    }
+    if (!matched) return false;
+  }
+  return true;
+}
+
+float SokobanNode::GoalDistanceEstimate(SokobanNode& goal_node) {
+  float h = 0;
+  for (size_t i = 0; i < boxes.size(); i++) {
```

Comment:
use range-based iteration: `for(const auto &box: boxes) {`

https://en.cppreference.com/w/cpp/language/range-for